console/console-init/ui/src/modules/device/components/DeviceFilter/DeviceFilter.tsx

```diff
@@ -77,95 +86,116 @@ export interface IDeviceFilter {
   };
   filterCriteria: IDeviceFilterCriteria[];
 }
+
 export interface IDeviceFilterProps {
   filter?: IDeviceFilter;
   setFilter?: (filter: IDeviceFilter) => void;
 }
+
 const DeviceFilter: React.FunctionComponent<IDeviceFilterProps> = (
   {
     // filter,
     // setFilter
   }
 ) => {
   const [filter, setFilter] = useState<IDeviceFilter>(setInitialFilter());
+  const [lastAppliedFilter, setLastAppliedFilter] = useState<IDeviceFilter[]>([
+    setInitialFilter()
+  ]);
+  const [isKebabOpen, setIsKebabOpen] = useState<boolean>(false);
   const typeOptions: ISelectOption[] = [
     {
       key: "direct",
       value: "direct",
-      label: "Directly connected",
-      isDisabled: false
+      label: "Directly connected"
     },
     {
       key: "gateway",
       value: "gateway",
-      label: "Using gateways",
-      isDisabled: false
+      label: "Using gateways"
     }
   ];
   const statusOptions: ISelectOption[] = [
     {
       key: "enabled",
       value: "enabled",
-      label: "Enabled",
-      isDisabled: false
+      label: "Enabled"
     },
     {
       key: "disabled",
       value: "disabled",
-      label: "Disabled",
-      isDisabled: false
+      label: "Disabled"
     }
   ];
   const timeOptions: ISelectOption[] = [
     {
       key: "hr",
       value: "hr",
-      label: "hr",
-      isDisabled: false
+      label: "hr"
     },
     {
       key: "min",
       value: "min",
-      label: "min",
-      isDisabled: false
+      label: "min"
     },
     {
       key: "sec",
       value: "sec",
-      label: "sec",
-      isDisabled: false
+      label: "sec"
     }
   ];
-
-  const setTime = (value: string, isStart: boolean, isFormat: boolean) => {
-    const filterObj = { ...filter };
-    if (isStart) {
-      let timeObj = filter.lastSeen.startTime;
-      if (isFormat) {
-        timeObj = { form: value, time: timeObj.time };
-      } else {
-        timeObj = { form: timeObj.form, time: value };
-      }
-      filterObj.lastSeen.startTime = timeObj;
-    } else {
-      let timeObj = filter.lastSeen.endTime;
-      if (isFormat) {
-        timeObj = { form: value, time: timeObj.time };
-      } else {
-        timeObj = { form: timeObj.form, time: value };
-      }
-      filterObj.lastSeen.endTime = timeObj;
-    }
+  const onClearFilter = () => {
+    setFilter(setInitialFilter());
+    setIsKebabOpen(false);
+  };
+  const onRedoFilter = () => {
+    // redoFilter();
+    const length = lastAppliedFilter.length;
+    const data = JSON.parse(
```

Comment:
not sure, why do you need this. first you stringify json then parse string in json. lastAppliedFilter is already in json format why do you need it. can you try to remove parse and stringify and check it's working or not.